### ISSUE TITLE
Add support for Gemini Pro v1.5

### ISSUE THREAD

**Issue by krschacht** [AllYourBot/hostedgpt#289]
The API for this just came out a week ago.

in backend/* we have separate ruby files for OpenAI and Anthropic. We need to model it off of this.

I don't know if there is a nice Ruby SDK for Google's API so we may have to do more manual HTTP work.

**Comment by papayalabs** [comment]
I think I could help with this feature.

**Comment by krschacht** [comment]
@papayalabs that would be great. This is library I’ve seen: https://github.com/gbaptista/gemini-ai But I haven’t looked extensively so there may be a better one.

Also, as you look at the backend/* feel free to sketch out an implementation plan if you want early feedback. Two things to note:

1) I have two different ways of stubbing out the test response. Copy the one that I use for openai (not the one for anthropic). It was complex enough that I refactored it first for openai to make sure it worked well, but it does.

2) I implemented function calling in openai but not yet in anthropic. I believe gemini supports function calling so it would be nice to add support for that.

I guess basically, model off of openai not anthropic, to the extent you notice differences.

**Comment by papayalabs** [comment]
Ok, great @krschacht. I already played with  https://github.com/gbaptista/gemini-ai. The only thing is that system instructions does not work correctly. I understand very well the implementation for backend/*in the first release v0.6 but how it is now I get lost. You guys call stream_next_conversation_message from AIBackend for OpenAI but in the case of Gemini I would create stream_next_conversation_message in AIBackend::Gemini. I am in the correct way?

**Comment by krschacht** [comment]
@papayalabs Yes, that's right. I did some refactoring to pull common things up into `ai_backend.rb`, so it's best to have that file open at the same time you're looking at `open_ai.rb`. I think the approach you should take is basically to make a copy of all the open_ai things, both here in ai_backend/* and within the test sub directory and just start implementing the gemini version of each one. Let me know if you get stuck.

**Comment by papayalabs** [comment]
For you to have an idea I already integrate Gemini and Ollama in the v.06 release. If I have any question I get back to you.
